### PR TITLE
Fix run-with-dates bugs and fix #403

### LINKED ISSUE #403: `subset_shape` with large input data
* xclim version: 0.15.0
* Python version: 3.8
* Operating System: ubuntu

### Description

Calling the `subset_shape` function on a large input dataset crashes when the dataset doesn't fit in memory. Not with a large polygon subset, but a large source dataset. So I want to subset a small portion of a large file.

### What I Did

I was using xclim through finch, the source data is the daily bccaqv2 datasets. 

### What I Received

```
finch         | INFO:PYWPS:Subsetting file 1 of 48 
finch         | Traceback (most recent call last):
finch         |   File "/opt/conda/envs/finch/lib/python3.8/site-packages/pywps/app/Process.py", line 243, in _run_process
finch         |     self.handler(wps_request, wps_response)  # the user must update the wps_response.
finch         |   File "/code/finch/processes/wps_base.py", line 36, in _handler_wrapper
finch         |     return self.wrapped_handler(request, response)
finch         |   File "/code/finch/processes/wps_ensemble_indices_polygon.py", line 74, in _handler
finch         |     return ensemble_common_handler(self, request, response, finch_subset_shape)
finch         |   File "/code/finch/processes/ensemble_utils.py", line 528, in ensemble_common_handler
finch         |     subsetted_files = subset_function(
finch         |   File "/code/finch/processes/subset.py", line 234, in finch_subset_shape
finch         |     process_threaded(_subset, netcdf_inputs)
finch         |   File "/code/finch/processes/utils.py", line 274, in process_threaded
finch         |     outputs = [function(r) for r in inputs]
finch         |   File "/code/finch/processes/utils.py", line 274, in <listcomp>
finch         |     outputs = [function(r) for r in inputs]
finch         |   File "/code/finch/processes/subset.py", line 219, in _subset
finch         |     subsetted = subset_shape(
finch         |   File "/opt/conda/envs/finch/lib/python3.8/site-packages/xclim/subset.py", line 208, in func_checker
finch         |     final = func(*formatted_args, **kwargs)
finch         |   File "/opt/conda/envs/finch/lib/python3.8/site-packages/xclim/subset.py", line 465, in subset_shape
finch         |     ds_copy = copy.deepcopy(ds)
finch         |   File "/opt/conda/envs/finch/lib/python3.8/copy.py", line 153, in deepcopy
finch         |     y = copier(memo)
finch         |   File "/opt/conda/envs/finch/lib/python3.8/site-packages/xarray/core/dataset.py", line 1181, in __deepcopy__
finch         |     return self.copy(deep=True)
finch         |   File "/opt/conda/envs/finch/lib/python3.8/site-packages/xarray/core/dataset.py", line 1073, in copy
finch         |     variables = {k: v.copy(deep=deep) for k, v in self._variables.items()}
finch         |   File "/opt/conda/envs/finch/lib/python3.8/site-packages/xarray/core/dataset.py", line 1073, in <dictcomp>
finch         |     variables = {k: v.copy(deep=deep) for k, v in self._variables.items()}
finch         |   File "/opt/conda/envs/finch/lib/python3.8/site-packages/xarray/core/variable.py", line 885, in copy
finch         |     data = np.array(data)
finch         |   File "/opt/conda/envs/finch/lib/python3.8/site-packages/xarray/core/indexing.py", line 677, in __array__
finch         |     self._ensure_cached()
finch         |   File "/opt/conda/envs/finch/lib/python3.8/site-packages/xarray/core/indexing.py", line 674, in _ensure_cached
finch         |     self.array = NumpyIndexingAdapter(np.asarray(self.array))
finch         |   File "/opt/conda/envs/finch/lib/python3.8/site-packages/numpy/core/_asarray.py", line 85, in asarray
finch         |     return array(a, dtype, copy=False, order=order)
finch         |   File "/opt/conda/envs/finch/lib/python3.8/site-packages/xarray/core/indexing.py", line 653, in __array__
finch         |     return np.asarray(self.array, dtype=dtype)
finch         |   File "/opt/conda/envs/finch/lib/python3.8/site-packages/numpy/core/_asarray.py", line 85, in asarray
finch         |     return array(a, dtype, copy=False, order=order)
finch         |   File "/opt/conda/envs/finch/lib/python3.8/site-packages/xarray/core/indexing.py", line 557, in __array__
finch         |     return np.asarray(array[self.key], dtype=None)
finch         |   File "/opt/conda/envs/finch/lib/python3.8/site-packages/numpy/core/_asarray.py", line 85, in asarray
finch         |     return array(a, dtype, copy=False, order=order)
finch         |   File "/opt/conda/envs/finch/lib/python3.8/site-packages/xarray/coding/variables.py", line 72, in __array__
finch         |     return self.func(self.array)
finch         |   File "/opt/conda/envs/finch/lib/python3.8/site-packages/xarray/coding/variables.py", line 218, in _scale_offset_decoding
finch         |     data = np.array(data, dtype=dtype, copy=True)
finch         |   File "/opt/conda/envs/finch/lib/python3.8/site-packages/xarray/coding/variables.py", line 72, in __array__
finch         |     return self.func(self.array)
finch         |   File "/opt/conda/envs/finch/lib/python3.8/site-packages/xarray/coding/variables.py", line 138, in _apply_mask
finch         |     data = np.asarray(data, dtype=dtype)
finch         |   File "/opt/conda/envs/finch/lib/python3.8/site-packages/numpy/core/_asarray.py", line 85, in asarray
finch         |     return array(a, dtype, copy=False, order=order)
finch         |   File "/opt/conda/envs/finch/lib/python3.8/site-packages/xarray/core/indexing.py", line 557, in __array__
finch         |     return np.asarray(array[self.key], dtype=None)
finch         |   File "/opt/conda/envs/finch/lib/python3.8/site-packages/xarray/backends/netCDF4_.py", line 72, in __getitem__
finch         |     return indexing.explicit_indexing_adapter(
finch         |   File "/opt/conda/envs/finch/lib/python3.8/site-packages/xarray/core/indexing.py", line 837, in explicit_indexing_adapter
finch         |     result = raw_indexing_method(raw_key.tuple)
finch         |   File "/opt/conda/envs/finch/lib/python3.8/site-packages/xarray/backends/netCDF4_.py", line 85, in _getitem
finch         |     array = getitem(original_array, key)
finch         |   File "netCDF4/_netCDF4.pyx", line 4389, in netCDF4._netCDF4.Variable.__getitem__
finch         | MemoryError: Unable to allocate 55.9 GiB for an array with shape (1068, 510, 55115) and data type int16
```

### PR DESCRIPTION
**Pull request by aulemahal**
### Pull Request Checklist:
- [x] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #403 and the issue in bird-house/finch#110.
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Documentation has been added / updated (for bug fixes / features)
- [x] HISTORY.rst has been updated (with summary of main changes)
- [ ] `bumpversion (minor / major / patch)` has been called on this branch
- [ ] Tags have been pushed (`git push --tags`)

* **What kind of change does this PR introduce?** 
This fixes bugs in the "run with date" methods of `run_length` that occurred when the date did not appear in the run (for example when resampling on frequencies smaller than a year). Also, I realized that the code structure inherited from when they were in the indices was not necessary, so it's now a bit simpler.

Also, removed the use of `deepcopy` in `subset_shape` as discussed in #403.

Finally, while I'm at it, replaced all uses of `resample().apply()` by `resample().map()`.

* **Does this PR introduce a breaking change?** 
No

* **Other information**:
There were no documentation changes cause there were no behaviour changes.

### CODE REVIEW

File: xclim/subset.py [hunk 1]
```diff
@@ -458,11 +457,10 @@ def subset_shape(
     """
     wgs84 = CRS("+proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs")
 
-    # TODO : edge case using polygon splitting decorator touches original ds when subsetting?
     if isinstance(ds, xarray.DataArray):
-        ds_copy = copy.deepcopy(ds._to_temp_dataset())
+        ds_copy = ds._to_temp_dataset()
```
Comment (Zeitsperre) [inline xclim/subset.py]: This is enough to prevent overflowing the memory?

File: tests/test_run_length.py [hunk 2]
```diff
@@ -37,15 +37,15 @@ def test_simple(self):
         )
         values[1:11] = 1
         da = xr.DataArray(values != 0, coords={"time": time}, dims="time")
-        lt = da.resample(time="M").apply(rl.longest_run_ufunc)
+        lt = da.resample(time="M").map(rl.longest_run_ufunc)
```
Comment (Zeitsperre) [inline tests/test_run_length.py]: What's the major difference? Is it the same method in the backend?

File: xclim/subset.py (hunk 1 above)
Comment (aulemahal) [reply xclim/subset.py]: Deepcopy was copying everything included the data, but  `ds.copy` is only creating a new Dataset with views to the same data. Same for `_to_temp_dataset()` in the case of DataArrays. When we reassign inplace further down, the Dataset replaces the view by a  new one, not touching the input data.

File: tests/test_run_length.py (hunk 2 above)
Comment (aulemahal) [reply tests/test_run_length.py]: I don't know! But there is a deprecation warning on `apply`, so why not.

Comment (Zeitsperre) [reply tests/test_run_length.py]: According to the Pandas documentation, `.map` is meant for general operations along the axis of interest but it looks like they have different behaviours depending on the Array/Series structure. Xarray probably follows this relatively closely. See: https://stackoverflow.com/questions/19798153/difference-between-map-applymap-and-apply-methods-in-pandas

File: xclim/subset.py (hunk 1 above)
Comment (Zeitsperre) [reply xclim/subset.py]: Useful to know. This could come in handy in other ways.

File: tests/test_run_length.py (hunk 2 above)
Comment (aulemahal) [reply tests/test_run_length.py]: I just did some quick research and it looks like the name "map" was thought to be more general and intuitive... http://xarray.pydata.org/en/stable/whats-new.html#id52 5th point.

**Pull request merged**